### ISSUE TITLE
Discussion(CI): Which jdks for testing?

### ISSUE THREAD

**Issue by MartinWitt** [INRIA/spoon#3691]
As we have a new JDK every half year and [java 16](https://openjdk.java.net/projects/jdk/16/) isn't that far away. We could starting a discussion what JDK versions are interesting for CI usage. Currently we use 8, 10 and 11. We could use all JDKs for testing, but the CI takes already really long for testing.
### travis limits:
A job gets canceled 
- When a job produces no log output for 10 minutes.
- When a job on a public repository takes longer than 50 minutes.
- When a job on a private repository takes longer than 120 minutes.

### JDKs
- 8 minimal version we support, important to test. 
- 9 breaking changes because module system
- 10 only `var` as interesting change, but shouldn't break anything.
- 11,12,13,14 nothing really interesting
- 15 latest/bleeding edge is always good to test

We could create a job for each JDK, or choose 3 (?) interesting for default CI and run with all JDKs only when some condition is met(if travis supports this, probably exploration needed). The condition could be : 
- A PR is marked as review
- A merge/commit is done to master
- A new version/snapshot gets released
### Other OS containers
Another discussion is, if we want to use windows containers in travis for more testing. Java on Windows should produce the same results as on unix, but sometimes tests are platform dependent without a good reason. For MacOS the situation is more complex, because travis changed the costs and MacOS seems not free anymore. Supporting MacOSin the CI, would either need free credits from travis or a macos docker container.
### Discussion 
Which JDK versions do you think are the best/most interesting for using in the CI?
Shall we extend the CI with windows runners?
Shall we split the CI in different complexity depending on the PR state?
Or do you want performance increases?

**Comment by monperrus** [comment]
> Which JDK versions do you think are the best/most interesting for using in the CI?

It would be great to replace Java 10 by Java 15

> Shall we extend the CI with windows runners?

We had that at one point with AppVeyor, but it was too unstable. We also had a Windows slave on the Spoon Jenkins but it did not work well. Any effort on this would be welcome.

**Comment by MartinWitt** [comment]
> It would be great to replace Java 10 by Java 15

Should be done with the new PR #3692 

> We had that at one point with AppVeyor, but it was too unstable. We also had a Windows slave on the Spoon Jenkins but it did not work well. Any effort on this would be welcome.

As there is a lot of bug backlog and I want to finish the architecture framework. I would add this to the end of backlog, but if somebody wants to research about travis go ahead. Maybe an issue helps finding someone with travis knowledge-

**Comment by slarse** [comment]
> Shall we extend the CI with windows runners?

[Travis CI has Windows support now](https://docs.travis-ci.com/user/reference/windows/#using-windows). The Spoon CI isn't terribly complicated, so it shouldn't be too hard to set up. Just add a new entry to the build matrix specifying `os: windows`, and a Windows-compatible script.

If that doesn't work, [GitHub Actions also has support for Windows](https://docs.github.com/en/free-pro-team@latest/actions/reference/specifications-for-github-hosted-runners#supported-software).

**Comment by slarse** [comment]
@MartinWitt I could try to set up a Windows build sometime in the coming week, I've done quite a lot of stuff with Travis over the years. Just from a quick glance it seems like a simple enough endeavor.

**Comment by MartinWitt** [comment]
> I could try to set up a Windows build sometime in the coming week

Would be awesome, because I have not much experience with travis. PR welcome.

> One thing to note: it will probably be a heck of a lot slower than the Linux build. Might be annoying for PRs.

Is it possible to run more tests/jobs for a PR marked as review? E.g. have a fast CI with jdk8 and latest for WIP PRs and the slower full run, when a PR reaches a state e.g. marked as review?

**Comment by slarse** [comment]
> > I could try to set up a Windows build sometime in the coming week
> 
> Would be awesome, because I have not much experience with travis. PR welcome. There could be some test failures.

Cool, I'll open a PR as soon as I've got some time on my hands.

> > One thing to note: it will probably be a heck of a lot slower than the Linux build. Might be annoying for PRs.
> 
> Is it possible to run more tests/jobs for a PR marked as review? E.g. have a fast CI with jdk8 and latest for WIP PRs and the slower full run, when a PR reaches a state e.g. marked as review?

Not that I know of. For conditional builds, [this is what's available](https://docs.travis-ci.com/user/conditions-v1#integration). Of course, one could let the build start, fetch the required information and then abort, but the initial startup time is typically pretty long for non-Linux builds. I'll see what I can do.

**Comment by slarse** [comment]
FYI I've gotten very much held up on other, more pressing issues (bugs) in Spoon, but I still intend to devote some time to the Windows builds in the near future. So unless someone else has time right now to do it, I will get it done eventually :)
